### ISSUE TITLE
CBM PET file type use - specifically REL

### ISSUE THREAD

**Issue by megatron-uk** [cc65/cc65#1602]
I've just been given a lovely PET 2001N-32 that I am attempting to write a text mode RPG for using CC65. 

At the moment I am testing in VICE using the 2001 PET model and ROMs.

One thing I am really wary of is the amount of data needed to fully represent a game like this - hundreds of screens, character lists, items, stats etc. With the amount of logic needed I intend to keep most game strings and game data on disk, loading as each "location" demands.

The direct access REL format would seem ideal for this to load single records without the overhead of either using cbm_load or cbm_k_getin reads through an entire file which will be horrendous, but how can we both create and read such a datafile with CC65?

Are those basic routines that create or open REL files somehow callable from within CC65, without wrappers like the other cbm_k functions?

**Comment by mrdudz** [comment]
I am not aware of an API for REL files. I am not aware of "Basic routines" that deal with them either however, its just OPEN/INPUT - implementing this using the cbm_k stuff shouldnt be too hard (and not create a lot of overhead either, especially if you do that in assembler)

**Comment by megatron-uk** [comment]
I suppose I should take a look at how the existing cmb_k_* functions are implemented on top of those lower level kernal calls and see how feasible it will be to build a set of new cbm_k_* methods for dealing with REL files.

I'm not an asm programmer (it's some 25+ years since I did any), but as long as it's not too hairy I guess it's feasible.

**Comment by mrdudz** [comment]
A lot of those are just direct calls to the kernal functions. to implement a REL file api you shouldnt create new cbm_k_ functions - since there are no kernal functions to deal with REL files at all.

**Comment by megatron-uk** [comment]
So where is the best place to look? The implementation of OPEN(with 'L'), RECORD and INPUT in the BASIC (v2? v4?) roms?

**Comment by polluks** [comment]
Yes, look at the BASIC implementation, RECORD are just CHROUT calls and so on.
You may translate https://www.c64-wiki.de/wiki/Relativdatei.

**Comment by greg-king5** [comment]
You can use `cbm_open()`, `cbm_read()`, and `cbm_write()`.  `cbm_record()` is the only one that's missing.  You can try this C implementation (untested).

`cmd` must be a logical file number of an already open DOS command/status channel.
`sec` must be the secondary address of your REL file.
`rec` must be the desired record number.
```c
#include <cbm.h>
#include <errno.h>

unsigned char __fastcall__ cbm_record(unsigned char cmd, unsigned char sec,
                                      unsigned int rec)
{
    static struct {
        const char doscom;
        unsigned char sec;
        unsigned int rec;
        unsigned char offset;
    } p = {
        'p',
        0,
        0,
        1
    };

    p.rec = rec;
    p.sec = sec;

    cbm_write(cmd, &p, sizeof p);
    return _oserror;
}
```

**Comment by megatron-uk** [comment]
Wow, that's a great start - many thanks for the pointers (that's a C pun, btw....) in the right direction.

My SD2PET-Future card has just arrived, so I can also start testing things out on real hardware.

**Comment by polluks** [comment]
Take care of the DOS bugs...
https://www.lyonlabs.org/commodore/onrequest/rel-file-bugs-commentary.txt

**Comment by mrdudz** [comment]
I'd be willing to merge a PR containing this library addition plus a test program that demonstrates how to use REL files :)
